### DISCUSSION TITLE
Prefer to use the Annotated version if possible. - why?

### DISCUSSION THREAD

**Discussion by AntiSol** [fastapi/fastapi#11910] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.post("/files/")
async def create_file(file: bytes = File()):
    return {"file_size": len(file)}


@app.post("/uploadfile/")
async def create_upload_file(file: UploadFile):
    return {"filename": file.filename}
```

### Description

The FastAPI docs say on many pages "Prefer to use the Annotated version if possible", but does not provide any reasoning behind this.

I have searched for a reason or advantage this extra boilerplate provides, but I haven't found any.

What advantage is there to using the annotated version that would cause me to prefer it?

### Operating System

Linux

### FastAPI Version

0.111.0

### Pydantic Version

2.8.2

### Python Version

3.11

**Comment by Kludex** [comment]
Type correctness, I guess. But where is it mentioned?

**Comment by AntiSol** [comment]
It's on many pages that provide examples. The one I was looking at is https://fastapi.tiangolo.com/tutorial/request-files/ . Choose the non-annotated version and you'll see:

![image](https://github.com/user-attachments/assets/c612df40-e2f6-446d-ac43-cfe8f9ecbdd6)

In what way is `file: Annotated[bytes, File()]` more correct than `file: bytes = File()`? To me the annotated version seems boilerplatey and unintuitive, whereas the non-annotated version seems straightforward and intuitive, it defines a type and a default and the two are separated clearly.

What advantage do I get from using the annotated version? Why do the dox make such a strong recommendation ("if possible")?
(<-- these are two distinct questions :) )

**Reply by Kludex** [reply, marked as answer]
> To me the annotated version seems boilerplatey and unintuitive

But as I said, it's the one that provides more type correctness.

`File` is a function that returns `Any`, so if you do `file: bytes = File()` you are saying that you `file` is a `bytes`, but you are passing `Any`. If you do `file: Annotated[bytes, File()]`, you are saying that bytes is the type, you don't have a default value, but in the context of FastAPI it's a file - since `File()` becomes a metadata for the type.

**Comment by AntiSol** [comment]
I see. So to answer my question, I would prefer the annotated version if I was using a type checker that threw a warning, presumably. And the advantage that I get is presumably "not annoying static analysis tools".

It seems to me though that an even more correct option would be to have the return value from File be more strongly typed, no? and rather than declaring the param as a `bytes` you'd use that class. Why does `File` return `Any` as opposed to some file class (probably wrapped in `Optional`)?

**Reply by ZHLONG321** [reply]
> I see. So to answer my question, I would prefer the annotated version if I was using a type checker that threw a warning, presumably. And the advantage that I get is presumably "not annoying static analysis tools".
> 
> It seems to me though that an even more correct option would be to have the return value from File be more strongly typed, no? and rather than declaring the param as a `bytes` you'd use that class. Why does `File` return `Any` as opposed to some file class (probably wrapped in `Optional`)?

I hit it.

If I write the function only for fastapi, It's OK.
```python
@app.get("/")
def foo(query: str = Query()):  # query: Annotated[str, Query()]
    print(query)
    return query.upper()
```

Maybe the type checker report warning like reportCallInDefaultInitializer `Function calls and mutable objects not allowed within parameter default value expression`, but I will supress it.

However, if `foo` need to be a normal function, `query` now has default value, call it without argument will not raise TypeError, all code before access the parameter will execute.
```python
>>> foo('bar')
bar
'BAR'
>>> foo()
annotation=str required=True alias='query' json_schema_extra={}
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 6, in foo
AttributeError: 'Query' object has no attribute 'upper'
```

Use Annotated will not pass default value to it, and will raise TypeError.
```python
... def foo(query: Annotated[str, Query()]):
...     print(query.upper())
...     return query.upper()
... 
>>> foo()
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: foo() missing 1 required positional argument: 'query'
```

Agian, I will not use Annotated if I write the function only for fastapi, no matter what others say.
Because it's complicated to change `foo: SomethingClass[GenericTypeVar] | SomethingClass2 = default_value` to `foo: Annotated[SomethingClass[GenericTypeVar] | SomethingClass2, ...] = default_value`

**Reply by AntiSol** [reply]
Interesting point!

But having thought about it, I don't really think it makes much sense to have `Query()` or `File()` as a default value for a function that isn't a fastapi endpoint, since both of these are things that come from the HTTP request - if you're passing a `Query` or `File` Object around then there must have been a request at some point, and that's the appropriate place to use `Query()` as the default.

IMHO the best solution for the type of situation you're describing @ZHLONG-CN would be to do something like this:

```python
@app.get("/")
def foo(query: str = Query()):  # query: Annotated[str, Query()]
  """ 
  API endpoint for foo function
  if you need to foo() directly from code (i.e not from a fastapi endpoint), use _real_foo() instead.
   """
  return _real_foo(query)

def _real_foo(query: str):
  """ heavy-lifter foo function that actually does the work, can be used from multiple places """ 
  print(query)
  return query.upper()

@app.get("/thing-and-foo")
def thing_and_foo(query: str = Query()):  # query: Annotated[str, Query()]
  """ 
  API endpoint that also uses foo() functionality
  if you need to foo() directly from code (i.e not from a fastapi endpoint), use _real_foo() instead.
   """
  query = thing(query)
  return _real_foo(query)
```

i.e `foo()` is only ever intended to be a fastapi endpoint and not called directly from code. It then passes the actual work to `_real_foo()`, which *is* intended to be called directly from code. 

this gives us:
```python
> _real_foo()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: _real_foo() missing 1 required positional argument: 'query'
```

> I will not use Annotated if I write the function only for fastapi, no matter what others say.
> Because it's complicated to change foo: SomethingClass[GenericTypeVar] | SomethingClass2 = default_value to foo: Annotated[SomethingClass[GenericTypeVar] | SomethingClass2, ...] = default_value

Yeah, agreed - As I said, imo the Annotated notation is needlessly verbose and boilerplatey, and provides no tangible advantage other than reducing false positives from defective type checkers. The correct solution is to fix or remove type checkers, and/or have stricter typing on the return from e.g `Query()` and `File()`.
